misk-testing/src/main/kotlin/misk/concurrent/FakeScheduledExecutorService.kt

```diff
@@ -46,39 +72,80 @@ class FakeScheduledExecutorService @Inject constructor(private val clock: Clock)
 
   private fun <V> scheduleTask(delay: Long, unit: TimeUnit, task: () -> V): ScheduledFuture<V> {
     val executeAt = clock.millis() + unit.toMillis(delay)
-    val future =
-        ScheduledFutureTask(executeAt, clock, task)
-    futures.put(executeAt, future)
+    val future = ScheduledFutureTask(executeAt, clock, task)
+    futures.add(future)
     return future
   }
 
   override fun scheduleAtFixedRate(
-    command: Runnable?,
+    command: Runnable,
     initialDelay: Long,
     period: Long,
-    unit: TimeUnit?
+    unit: TimeUnit
   ): ScheduledFuture<*> {
-    TODO("not implemented")
+    check(period > 0) { "Period ($period) must be greater than 0" }
+    return scheduleWithFixedDelay(command, initialDelay, period, unit)
   }
 
   override fun scheduleWithFixedDelay(
-    command: Runnable?,
+    command: Runnable,
     initialDelay: Long,
     delay: Long,
-    unit: TimeUnit?
+    unit: TimeUnit
   ): ScheduledFuture<*> {
-    TODO("not implemented")
+    check(delay > 0) { "Delay ($delay) must be greater than 0" }
+
+    val executeAt = clock.millis() + unit.toMillis(initialDelay)
+    val everyMillis = unit.toMillis(delay)
+
+    val future = ScheduledFutureTask(executeAt, everyMillis, clock) { command.run() }
+    futures.add(future)
+    return future
   }
 
-  class ScheduledFutureTask<V>(
-    val executeAt: Long,
+  inner class ScheduledFutureTask<V>(
+    var executeAt: Long,
+    private val fixedDelay: Long,
     val clock: Clock,
-    task: () -> V
+    val task: () -> V
   ) : FutureTask<V>(task), ScheduledFuture<V> {
+    constructor(
+      executeAt: Long,
+      clock: Clock,
+      task: () -> V
+    ) : this(executeAt, 0L, clock, task)
+
+    private var cancelled = false
+
     override fun compareTo(other: Delayed): Int =
         getDelay(TimeUnit.MILLISECONDS).compareTo(other.getDelay(TimeUnit.MILLISECONDS))
 
     override fun getDelay(unit: TimeUnit) =
         unit.convert(clock.millis() - executeAt, TimeUnit.MILLISECONDS)
+
+    override fun isCancelled(): Boolean {
+      return cancelled
+    }
+
+    val isRepeated: Boolean get() = fixedDelay > 0
+
+    override fun run() {
+      if (isCancelled) throw CancellationException("FutureTask was cancelled")
+      super.run()
+    }
+
+    public override fun runAndReset(): Boolean {
+      return super.runAndReset()
+    }
```

Comment:
This seems superfluous?

Reply:
I would agree, but unfortunately it seems that `runAndReset()` needs to be overridden to make it a public method, so it may be used by the enclosing class.